### ISSUE TITLE
Machine Translation with Recurrent Neural Networks

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7949]
Machine Translation with Recurrent Neural Networks<br>
<br>
https://ift.tt/3xJR867<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<div>
<pre><span>class</span> <span>Seq2seq</span><span>(</span><span>nn</span><span>.</span><span>Module</span><span>):</span>
    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>embedding_src</span><span>,</span> <span>embedding_tgt</span><span>,</span> <span>h_dim</span><span>,</span> <span>num_layers</span><span>,</span> <span>dropout_p</span><span>,</span> <span>bi</span><span>,</span> <span>tokens_bos_eos_pad_unk</span><span>=</span><span>[</span><span>0</span><span>,</span><span>1</span><span>,</span><span>2</span><span>,</span><span>3</span><span>]):</span>
        <span>super</span><span>(</span><span>Seq2seq</span><span>,</span> <span>self</span><span>)</span><span>.</span><span>__init__</span><span>()</span>
        <span># Store hyperparameters</span>
        <span>self</span><span>.</span><span>h_dim</span> <span>=</span> <span>h_dim</span>
        <span>self</span><span>.</span><span>vocab_size_tgt</span><span>,</span> <span>self</span><span>.</span><span>emb_dim_tgt</span> <span>=</span> <span>embedding_tgt</span><span>.</span><span>size</span><span>()</span>
        <span>self</span><span>.</span><span>bos_token</span><span>,</span> <span>self</span><span>.</span><span>eos_token</span><span>,</span> <span>self</span><span>.</span><span>pad_token</span><span>,</span> <span>self</span><span>.</span><span>unk_token</span> <span>=</span> <span>tokens_bos_eos_pad_unk</span>

        <span># Create encoder, decoder, attention</span>
        <span>self</span><span>.</span><span>encoder</span> <span>=</span> <span>EncoderLSTM</span><span>(</span><span>embedding_src</span><span>,</span> <span>h_dim</span><span>,</span> <span>num_layers</span><span>,</span> <span>dropout_p</span><span>=</span><span>dropout_p</span><span>,</span> <span>bidirectional</span><span>=</span><span>bi</span><span>)</span>
        <span>self</span><span>.</span><span>decoder</span> <span>=</span> <span>DecoderLSTM</span><span>(</span><span>embedding_tgt</span><span>,</span> <span>h_dim</span><span>,</span> <span>num_layers</span> <span>*</span> <span>2</span> <span>if</span> <span>bi</span> <span>else</span> <span>num_layers</span><span>,</span> <span>dropout_p</span><span>=</span><span>dropout_p</span><span>)</span>
        <span>self</span><span>.</span><span>attention</span> <span>=</span> <span>Attention</span><span>(</span><span>pad_token</span><span>=</span><span>self</span><span>.</span><span>pad_token</span><span>,</span> <span>bidirectional</span><span>=</span><span>bi</span><span>,</span> <span>h_dim</span><span>=</span><span>self</span><span>.</span><span>h_dim</span><span>)</span>

        <span># Create linear layers to combine context and hidden state</span>
        <span>self</span><span>.</span><span>linear1</span> <span>=</span> <span>nn</span><span>.</span><span>Linear</span><span>(</span><span>2</span> <span>*</span> <span>self</span><span>.</span><span>h_dim</span><span>,</span> <span>self</span><span>.</span><span>emb_dim_tgt</span><span>)</span>
        <span>self</span><span>.</span><span>tanh</span> <span>=</span> <span>nn</span><span>.</span><span>Tanh</span><span>()</span>
        <span>self</span><span>.</span><span>dropout</span> <span>=</span> <span>nn</span><span>.</span><span>Dropout</span><span>(</span><span>dropout_p</span><span>)</span>
        <span>self</span><span>.</span><span>linear2</span> <span>=</span> <span>nn</span><span>.</span><span>Linear</span><span>(</span><span>self</span><span>.</span><span>emb_dim_tgt</span><span>,</span> <span>self</span><span>.</span><span>vocab_size_tgt</span><span>)</span>
        
        <span># Share weights between decoder embedding and output </span>
        <span>if</span> <span>self</span><span>.</span><span>decoder</span><span>.</span><span>embedding</span><span>.</span><span>weight</span><span>.</span><span>size</span><span>()</span> <span>==</span> <span>self</span><span>.</span><span>linear2</span><span>.</span><span>weight</span><span>.</span><span>size</span><span>():</span>
            <span>self</span><span>.</span><span>linear2</span><span>.</span><span>weight</span> <span>=</span> <span>self</span><span>.</span><span>decoder</span><span>.</span><span>embedding</span><span>.</span><span>weight</span>

    <span>def</span> <span>forward</span><span>(</span><span>self</span><span>,</span> <span>src</span><span>,</span> <span>tgt</span><span>):</span>
        <span>if</span> <span>use_gpu</span><span>:</span> <span>src</span> <span>=</span> <span>src</span><span>.</span><span>cuda</span><span>()</span>
        
        <span># Encode</span>
        <span>out_e</span><span>,</span> <span>final_e</span> <span>=</span> <span>self</span><span>.</span><span>encoder</span><span>(</span><span>src</span><span>)</span>
        
        <span># Decode</span>
        <span>out_d</span><span>,</span> <span>final_d</span> <span>=</span> <span>self</span><span>.</span><span>decoder</span><span>(</span><span>tgt</span><span>,</span> <span>final_e</span><span>)</span>
        
        <span># Attend</span>
        <span>context</span> <span>=</span> <span>self</span><span>.</span><span>attention</span><span>(</span><span>src</span><span>,</span> <span>out_e</span><span>,</span> <span>out_d</span><span>)</span>
        <span>out_cat</span> <span>=</span> <span>torch</span><span>.</span><span>cat</span><span>((</span><span>out_d</span><span>,</span> <span>context</span><span>),</span> <span>dim</span><span>=</span><span>2</span><span>)</span> 
        
        <span># Predict (returns probabilities)</span>
        <span>x</span> <span>=</span> <span>self</span><span>.</span><span>linear1</span><span>(</span><span>out_cat</span><span>)</span>
        <span>x</span> <span>=</span> <span>self</span><span>.</span><span>dropout</span><span>(</span><span>self</span><span>.</span><span>tanh</span><span>(</span><span>x</span><span>))</span>
        <span>x</span> <span>=</span> <span>self</span><span>.</span><span>linear2</span><span>(</span><span>x</span><span>)</span>
        <span>return</span> <span>x</span>

    <span>def</span> <span>predict</span><span>(</span><span>self</span><span>,</span> <span>src</span><span>,</span> <span>beam_size</span><span>=</span><span>1</span><span>):</span> 
        <span>'''Predict top 1 sentence using beam search. Note that beam_size=1 is greedy search.'''</span>
        <span>beam_outputs</span> <span>=</span> <span>self</span><span>.</span><span>beam_search</span><span>(</span><span>src</span><span>,</span> <span>beam_size</span><span>,</span> <span>max_len</span><span>=</span><span>30</span><span>)</span> <span># returns top beam_size options (as list of tuples)</span>
        <span>top1</span> <span>=</span> <span>beam_outputs</span><span>[</span><span>0</span><span>][</span><span>1</span><span>]</span> <span># a list of word indices (as ints)</span>
        <span>return</span> <span>top1</span>

    <span>def</span> <span>beam_search</span><span>(</span><span>self</span><span>,</span> <span>src</span><span>,</span> <span>beam_size</span><span>,</span> <span>max_len</span><span>,</span> <span>remove_tokens</span><span>=</span><span>[]):</span>
        <span>'''Returns top beam_size sentences using beam search. Works only when src has batch size 1.'''</span>
        <span>if</span> <span>use_gpu</span><span>:</span> <span>src</span> <span>=</span> <span>src</span><span>.</span><span>cuda</span><span>()</span>
        
        <span># Encode</span>
        <span>outputs_e</span><span>,</span> <span>states</span> <span>=</span> <span>self</span><span>.</span><span>encoder</span><span>(</span><span>src</span><span>)</span> <span># batch size = 1</span>
        
        <span># Start with '&lt;s&gt;'</span>
        <span>init_lprob</span> <span>=</span> <span>-</span><span>1e10</span>
        <span>init_sent</span> <span>=</span> <span>[</span><span>self</span><span>.</span><span>bos_token</span><span>]</span>
        <span>best_options</span> <span>=</span> <span>[(</span><span>init_lprob</span><span>,</span> <span>init_sent</span><span>,</span> <span>states</span><span>)]</span> <span># beam</span>
        
        <span># Beam search</span>
        <span>k</span> <span>=</span> <span>beam_size</span> <span># store best k options</span>
        <span>for</span> <span>length</span> <span>in</span> <span>range</span><span>(</span><span>max_len</span><span>):</span> <span># maximum target length</span>
            <span>options</span> <span>=</span> <span>[]</span> <span># candidates </span>
            <span>for</span> <span>lprob</span><span>,</span> <span>sentence</span><span>,</span> <span>current_state</span> <span>in</span> <span>best_options</span><span>:</span>
                <span># Prepare last word</span>
                <span>last_word</span> <span>=</span> <span>sentence</span><span>[</span><span>-</span><span>1</span><span>]</span>
                <span>if</span> <span>last_word</span> <span>!=</span> <span>self</span><span>.</span><span>eos_token</span><span>:</span>
                    <span>last_word_input</span> <span>=</span> <span>torch</span><span>.</span><span>LongTensor</span><span>([</span><span>last_word</span><span>])</span><span>.</span><span>view</span><span>(</span><span>1</span><span>,</span><span>1</span><span>)</span>
                    <span>if</span> <span>use_gpu</span><span>:</span> <span>last_word_input</span> <span>=</span> <span>last_word_input</span><span>.</span><span>cuda</span><span>()</span>
                    <span># Decode</span>
                    <span>outputs_d</span><span>,</span> <span>new_state</span> <span>=</span> <span>self</span><span>.</span><span>decoder</span><span>(</span><span>last_word_input</span><span>,</span> <span>current_state</span><span>)</span>
                    <span># Attend</span>
                    <span>context</span> <span>=</span> <span>self</span><span>.</span><span>attention</span><span>(</span><span>src</span><span>,</span> <span>outputs_e</span><span>,</span> <span>outputs_d</span><span>)</span>
                    <span>out_cat</span> <span>=</span> <span>torch</span><span>.</span><span>cat</span><span>((</span><span>outputs_d</span><span>,</span> <span>context</span><span>),</span> <span>dim</span><span>=</span><span>2</span><span>)</span>
                    <span>x</span> <span>=</span> <span>self</span><span>.</span><span>linear1</span><span>(</span><span>out_cat</span><span>)</span>
                    <span>x</span> <span>=</span> <span>self</span><span>.</span><span>dropout</span><span>(</span><span>self</span><span>.</span><span>tanh</span><span>(</span><span>x</span><span>))</span>
                    <span>x</span> <span>=</span> <span>self</span><span>.</span><span>linear2</span><span>(</span><span>x</span><span>)</span>
                    <span>x</span> <span>=</span> <span>x</span><span>.</span><span>squeeze</span><span>()</span><span>.</span><span>data</span><span>.</span><span>clone</span><span>()</span>
                    <span># Block predictions of tokens in remove_tokens</span>
                    <span>for</span> <span>t</span> <span>in</span> <span>remove_tokens</span><span>:</span> <span>x</span><span>[</span><span>t</span><span>]</span> <span>=</span> <span>-</span><span>10e10</span>
                    <span>lprobs</span> <span>=</span> <span>torch</span><span>.</span><span>log</span><span>(</span><span>x</span><span>.</span><span>exp</span><span>()</span> <span>/</span> <span>x</span><span>.</span><span>exp</span><span>()</span><span>.</span><span>sum</span><span>())</span> <span># log softmax</span>
                    <span># Add top k candidates to options list for next word</span>
                    <span>for</span> <span>index</span> <span>in</span> <span>torch</span><span>.</span><span>topk</span><span>(</span><span>lprobs</span><span>,</span> <span>k</span><span>)[</span><span>1</span><span>]:</span> 
                        <span>option</span> <span>=</span> <span>(</span><span>float</span><span>(</span><span>lprobs</span><span>[</span><span>index</span><span>])</span> <span>+</span> <span>lprob</span><span>,</span> <span>sentence</span> <span>+</span> <span>[</span><span>index</span><span>],</span> <span>new_state</span><span>)</span> 
                        <span>options</span><span>.</span><span>append</span><span>(</span><span>option</span><span>)</span>
                <span>else</span><span>:</span> <span># keep sentences ending in '&lt;/s&gt;' as candidates</span>
                    <span>options</span><span>.</span><span>append</span><span>((</span><span>lprob</span><span>,</span> <span>sentence</span><span>,</span> <span>current_state</span><span>))</span>
            <span>options</span><span>.</span><span>sort</span><span>(</span><span>key</span> <span>=</span> <span>lambda</span> <span>x</span><span>:</span> <span>x</span><span>[</span><span>0</span><span>],</span> <span>reverse</span><span>=</span><span>True</span><span>)</span> <span># sort by lprob</span>
            <span>best_options</span> <span>=</span> <span>options</span><span>[:</span><span>k</span><span>]</span> <span># place top candidates in beam</span>
        <span>best_options</span><span>.</span><span>sort</span><span>(</span><span>key</span> <span>=</span> <span>lambda</span> <span>x</span><span>:</span> <span>x</span><span>[</span><span>0</span><span>],</span> <span>reverse</span><span>=</span><span>True</span><span>)</span>
        <span>return</span> <span>best_options</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live.<br>
<br>
June 18, 2021 at 08:19PM
